tools/ja-rule/JaRuleWidgetImpl.cpp

```diff
@@ -0,0 +1,251 @@
+  OLA_INFO << "Full discovery triggered";
+  m_discovery_agent.StartFullDiscovery(
+      NewSingleCallback(this, &JaRuleWidgetImpl::DiscoveryComplete,
+      callback));
+}
+
+void JaRuleWidgetImpl::RunIncrementalDiscovery(RDMDiscoveryCallback *callback) {
+  OLA_INFO << "Incremental discovery triggered";
+  m_discovery_agent.StartIncrementalDiscovery(
+      NewSingleCallback(this, &JaRuleWidgetImpl::DiscoveryComplete,
+      callback));
+}
+
+
+void JaRuleWidgetImpl::SendRDMRequest(const RDMRequest *request,
+                                      ola::rdm::RDMCallback *on_complete) {
+  if (!CheckForDevice()) {
+    const std::vector<std::string> packets;
+    on_complete->Run(ola::rdm::RDM_FAILED_TO_SEND, NULL, packets);
+    return;
+  }
+
+  m_rdm_callback = on_complete;
+
+  unsigned int rdm_length = RDMCommandSerializer::RequiredSize(*request);
+  uint8_t data[rdm_length];
+  RDMCommandSerializer::Pack(*request, data, &rdm_length);
+  m_device->SendMessage(JaRuleEndpoint::RDM_REQUEST, data, rdm_length);
+}
+
+void JaRuleWidgetImpl::MuteDevice(const UID &target,
+                                  MuteDeviceCallback *mute_complete) {
+  if (!CheckForDevice()) {
+    mute_complete->Run(false);
+    return;
+  }
+
+  auto_ptr<RDMRequest> request(
+      ola::rdm::NewMuteRequest(m_our_uid, target,
+                               m_transaction_number.Next()));
+
+  unsigned int rdm_length = RDMCommandSerializer::RequiredSize(*request);
+  uint8_t data[rdm_length];
+  RDMCommandSerializer::Pack(*request, data, &rdm_length);
+  m_device->SendMessage(JaRuleEndpoint::RDM_REQUEST, data, rdm_length);
+
+  m_mute_callback = mute_complete;
+}
+
+void JaRuleWidgetImpl::UnMuteAll(UnMuteDeviceCallback *unmute_complete) {
+  if (!CheckForDevice()) {
+    unmute_complete->Run();
+    return;
+  }
+
+  auto_ptr<RDMRequest> request(
+      ola::rdm::NewUnMuteRequest(m_our_uid, UID::AllDevices(),
+                                 m_transaction_number.Next()));
+
+  unsigned int rdm_length = RDMCommandSerializer::RequiredSize(*request);
+  uint8_t data[rdm_length];
+  RDMCommandSerializer::Pack(*request, data, &rdm_length);
+  m_device->SendMessage(JaRuleEndpoint::RDM_REQUEST, data, rdm_length);
+
+  m_unmute_callback = unmute_complete;
+}
+
+void JaRuleWidgetImpl::Branch(const UID &lower,
+                              const UID &upper,
+                              BranchCallback *branch_complete) {
+  if (!CheckForDevice()) {
+    branch_complete->Run(NULL, 0);
+    return;
+  }
+
+  auto_ptr<RDMRequest> request(
+      ola::rdm::NewDiscoveryUniqueBranchRequest(m_our_uid, lower, upper,
+                                                m_transaction_number.Next()));
+  unsigned int rdm_length = RDMCommandSerializer::RequiredSize(*request);
+  uint8_t data[rdm_length];
+  RDMCommandSerializer::Pack(*request, data, &rdm_length);
+  OLA_INFO << "Sending " << rdm_length << " RDM command: " << lower << " - "
```

Comment:
Can we add the word byte here so it reads better? Perhaps also add a note about DUB, or at least clarify what the two UIDs are.